### ISSUE TITLE
Misc. bug: Router - Queue model loading requests instead of unloading models with active queries

### ISSUE THREAD

**Issue by jmig1109** [ggml-org/llama.cpp#21678]
### Name and Version

llama-server --version
version: 8736 (c8ac02fa1)
built with GNU 13.3.0 for Linux x86_64

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server

### Command line

```shell
CUDA_VISIBLE_DEVICES=1,0,2 CUDA_SCALE_LAUNCH_QUEUES=4x llama-server   --models-preset /mnt/fast1/gguf-models/models-build.ini   -t 6   --host 0.0.0.0   --port 8080   --models-max 1   -fa auto   --log-colors on   -cram 32768   --slot-save-path /mnt/fast1/kv-cache-storage   --cache-prompt   --jinja
```

### Problem description & steps to reproduce

Problem description
When models_max is reached in the router server, the LRU unloading logic (unload_lru) removes the least recently used model to make room for a new one without checking if that model is currently processing active requests. This causes ongoing queries to be interrupted when a request for a different model triggers an immediate unload of the active model.
Steps to reproduce
1. Start the router server with models_max=1 and at least two available models (Model A and Model B).
2. Send a long-running query (e.g., a large generation) to Model A.
3. While Model A is still generating, send a query to Model B.
4. Observe that the server immediately unloads Model A to load Model B, interrupting the first query.

### First Bad Commit

Unknown

### Relevant log output

<details>
<summary>Logs</summary>

```console
[56021] main: model loaded
[56021] main: server is listening on http://127.0.0.1:56021
[56021] main: starting the main loop...
[56021] cmd_child_to_router:ready
srv  proxy_reques: proxying request to model Gemma-4-Pro on port 56021
[56021] srv  update_slots: all slots are idle
[56021] srv    operator(): child server monitoring thread started, waiting for EOF on stdin...
[56021] srv  params_from_: Chat format: peg-gemma4
[56021] slot get_availabl: id  1 | task -1 | selected slot by LRU, t_last = -1
[56021] srv  get_availabl: updating prompt cache
[56021] srv          load:  - looking for better prompt, base f_keep = -1.000, sim = 0.000
[56021] srv        update:  - cache state: 0 prompts, 0.000 MiB (limits: 32768.000 MiB, 262144 tokens, 34359738368 est)
[56021] srv  get_availabl: prompt cache update took 0.00 ms
[56021] slot launch_slot_: id  1 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> ?temp-ext -> dist 
[56021] slot launch_slot_: id  1 | task 0 | processing task, is_child = 0
[56021] slot update_slots: id  1 | task 0 | new prompt, n_ctx_slot = 262144, n_keep = 0, task.n_tokens = 82470
[56021] slot update_slots: id  1 | task 0 | n_tokens = 0, memory_seq_rm [0, end)
[56021] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 8192, batch.n_tokens = 8192, progress = 0.099333
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
srv  proxy_reques: proxying request to model Gemma-4-Pro on port 56021
[56021] slot update_slots: id  1 | task 0 | n_tokens = 8192, memory_seq_rm [8192, end)
[56021] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 16384, batch.n_tokens = 8192, progress = 0.198666
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
[56021] slot update_slots: id  1 | task 0 | n_tokens = 16384, memory_seq_rm [16384, end)
[56021] slot update_slots: id  1 | task 0 | 16384 tokens since last checkpoint at 0, creating new checkpoint during processing at position 24576
[56021] slot update_slots: id  1 | task 0 | prompt processing progress, n_tokens = 24576, batch.n_tokens = 8192, progress = 0.297999
[56021] slot update_slots: id  1 | task 0 | created context checkpoint 1 of 8 (pos_min = 13824, pos_max = 16383, n_tokens = 16384, size = 2000.031 MiB)
srv    unload_lru: models_max limit reached, removing LRU name=Gemma-4-Pro
srv        unload: stopping model instance name=Gemma-4-Pro
srv    operator(): stopping model instance name=Gemma-4-Pro
[56021] srv    operator(): exit command received, exiting...
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
[56021] srv    operator(): operator(): cleaning up before exit...
srv  log_server_r: done request: GET / 172.23.0.2 200
srv  log_server_r: done request: GET / 192.168.129.119 200
srv    operator(): force-killing model instance name=Gemma-4-Pro after 10 seconds timeout
srv    operator(): http client error: Failed to read connection
srv  log_server_r: done request: POST /v1/chat/completions 192.168.129.119 500
srv  log_server_r: done request: GET / 172.23.0.2 200
srv    operator(): instance name=Gemma-4-Pro exited with status 1
srv          load: spawning server instance with name=Gemma-4-Flash on port 41753
```
</details>

**Comment by tha80** [comment]
I can confirm that the reported issue exists as I am also having this problem when a request for model A is running while a second request, but this time for model B, hits the router. The active request is cancelled, the client receives an error and the model is unloaded before model B is then started.

In my setup, where all of my clients are retrying failed requests directly, this can easily lead to a situation where llama.cpp is stuck in a loop of unloading and loading modules without actually processing any requests anymore.

I guess the best solution would be to make the router wait until there is no active request anymore, before actually unloading it.

**Comment by jmig1109** [comment]
> I guess the best solution would be to make the router wait until there is no active request anymore, before actually unloading it.

Either that or a flag to decide how to handle unloading in such a situation - immediate unload/reload; execute queued tasks sequentially.

Models-max 1 and the router server are a godsend for everyone with a small setup and limited VRAM/RAM but there are a couple of frustrations like this one.

**Comment by Haervwe** [comment]
I can also confirm this is an issue, the server needs a queue of sorts for requests to diferent models. im using different presets of the same model for different tasks on hermers agents + hindsight and there are cases where i hit a loop of bassically loading and unloading in a loop until one of the processess give up.

**Comment by loopy321** [comment]
Confirmed issue for me too.  llama-server (version 9297) router mode with --models-max 1 evicts/kills an active or still-loading model when a request for a different model arrives, instead of queueing the second request until the first model’s active request completes. This violates expected FIFO/single-flight behavior and causes interrupted generations despite --parallel 1.

**Comment by fitdev** [comment]
Yeah this is an issue for me too.

**Comment by jmig1109** [comment]
Thanks @loopy321 for creating a PR for this! let's hope it gets solved

**Comment by 204504bySE** [comment]
I wrote a workaround reverse proxy.
https://github.com/204504bySE/LlamaWorkaround

**Comment by XBold** [comment]
Hi all, still having the same issue.
In the future will be implemented the feature of some sort of queue? It would be really helpful.
Thanks to all! :D

**Comment by ngxson** [comment]
I implemented a complete LRU scheduler in https://github.com/ggml-org/llama.cpp/pull/26572 , please give it a try

**Comment by jmig1109** [comment]
Thanks a lot! checked the pr out, this fixes the issue for me.

**Comment by loopy321** [comment]
Working for me.  Thanks
